### PR TITLE
[improvement] Allow setting additional docker run args with arguments property

### PR DESCRIPTION
**Pull request by christopherbryanboyd**
## Before this PR
dockerRun is a very useful plugin, but it has limitations on how it can be used, specifically in regard to unsupported arguments.

## After this PR
This PR adds an extension property to dockerRun to allow setting custom arguments for docker run, in this form:
```
dockerRun {
    arguments '-P', 'w=/test'
}
```
This PR also includes a test for this functionality.

### CODE REVIEW

File: src/main/groovy/com/palantir/gradle/docker/DockerRunExtension.groovy [hunk 1]
```diff
@@ -93,14 +94,22 @@ class DockerRunExtension {
     private void setEnvSingle(String key, String value) {
         this.env.put(checkNotNull(key, "key"), checkNotNull(value, "value"))
     }
-
-    public void env(Map<String,String> env) {
-        this.env = ImmutableMap.copyOf(env)
-    }
+	
+	public void env(Map<String,String> env) {
```
Comment (pkoenig10) [inline src/main/groovy/com/palantir/gradle/docker/DockerRunExtension.groovy]: Please fix the indentation in this file

File: src/main/groovy/com/palantir/gradle/docker/DockerRunPlugin.groovy [hunk 2]
```diff
@@ -123,7 +123,10 @@ class DockerRunPlugin implements Plugin<Project> {
                     args.add("${localFile.absolutePath}:${volume.value}")
                 }
                 args.addAll(ext.env.collect{ k, v -> ['-e', "${k}=${v}"] }.flatten())
-                args.addAll(['--name', ext.name, ext.image])
+				for (String argument : ext.arguments) {
+					args.add(argument)
+				}
+				args.addAll(['--name', ext.name, ext.image])
```
Comment (pkoenig10) [inline src/main/groovy/com/palantir/gradle/docker/DockerRunPlugin.groovy]: nit: add this argument before adding the custom arguments

**Review by pkoenig10 — COMMENTED** [review]
This seems like a reasonable way to support additional flags without increasing the complexity of this plugin. I assume this supersedes #244 and #245?

File: src/main/groovy/com/palantir/gradle/docker/DockerRunExtension.groovy (hunk 1 above)
Comment (christopherbryanboyd) [reply src/main/groovy/com/palantir/gradle/docker/DockerRunExtension.groovy]: Fixed, sorry it was a spaces - tabs issue.

**Comment by christopherbryanboyd** [comment]
Hi @pkoenig10 , this does supersede the other 2 PR's, but maybe those would be helpful as well to users who don't want to specify the arguments directly, or want a more idiomatic / natural way of specifying them?

File: src/main/groovy/com/palantir/gradle/docker/DockerRunPlugin.groovy (hunk 2 above)
Comment (christopherbryanboyd) [reply src/main/groovy/com/palantir/gradle/docker/DockerRunPlugin.groovy]: Hi @pkoenig10 , I updated this section to specify `--name` before custom arguments, but `ext.image` must still come after the custom arguments, as per the docker run manual, the name of the image should come after all of the flags, but before the actual command being executed.

**Comment by pkoenig10** [comment]
I would prefer to avoid adding additional first-class options unless they are a features used/required by almost every docker image build. See https://github.com/palantir/gradle-docker/pull/186#issuecomment-400864547.

This PR makes it easy to add custom options without incurring the cost of scope creep in this plugin.

**Review by pkoenig10 — APPROVED** [review]

**Comment by christopherbryanboyd** [comment]
Sounds good to me!

File: src/test/groovy/com/palantir/gradle/docker/DockerRunPluginTests.groovy [hunk 3]
```diff
@@ -153,6 +153,42 @@ class DockerRunPluginTests extends AbstractPluginTest {
...
+		when:
+	   BuildResult buildResult = with('docker', 'dockerRemoveContainer', 'dockerRun', 'dockerRunStatus').build()
+
+		then:
+	   buildResult.task(':dockerRemoveContainer').outcome == TaskOutcome.SUCCESS
+
+		buildResult.task(':dockerRun').outcome == TaskOutcome.SUCCESS
+	   buildResult.output =~ /(?m)\/test/
+
+		buildResult.task(':dockerRunStatus').outcome == TaskOutcome.SUCCESS
```
Comment (pkoenig10) [inline src/test/groovy/com/palantir/gradle/docker/DockerRunPluginTests.groovy]: nit: fix the indentation in this tests

File: src/main/groovy/com/palantir/gradle/docker/DockerRunExtension.groovy [hunk 4]
```diff
@@ -93,14 +94,22 @@ class DockerRunExtension {
     private void setEnvSingle(String key, String value) {
         this.env.put(checkNotNull(key, "key"), checkNotNull(value, "value"))
     }
-
+
```
Comment (pkoenig10) [inline src/main/groovy/com/palantir/gradle/docker/DockerRunExtension.groovy]: nit: remove whitespace

**Review by pkoenig10 — COMMENTED** [review]
Couple minor nits then this is good to merge

**Comment by christopherbryanboyd** [comment]
Hi @pkoenig10 , I have implemented the requested changes / fixes to the whitespace and indentation.

**Comment by mngo87** [comment]
+1 to this. Would love to use dockerRun plugin, but am unable to until this PR is merged in

**Comment by ferozco** [comment]
👍

**Comment by christopherbryanboyd** [comment]
@dansanduleac Do you know why this PR did not get merged?

**Comment by ferozco** [comment]
Looks like some automation did not work as intended. Thanks for the contribution @christopherbryanboyd! I've gone ahead and released the change as part of 0.25.0

**Comment by UweNachname** [comment]
Thank you christopherbryanboyd. This is exactly what I needed and it solved my problem. Very great !!!

**Pull request merged**